### DISCUSSION TITLE
can acl works while jwt expired?

### DISCUSSION THREAD

**Discussion by hashforu** [emqx/emqx#9067] [Q&A] [answered]
i'm not clearly how jwt plugin works,could anybody explain to me.
i mean,  client got token from my external auth server, then login with it.
the token contains filed 'iat', and emqx verify passed, allow to connect.
the token also contaions acl, if the token expired, i need the client keep in connected, and acl works(in other words, **acl have a long lifetime than token, it should works the whole time, except connection lost**).

how acl works?  check the jwt fileds every pub/pub action?  will verify the expired filed the same time? 
and ,will it cached after first authrization? 

thanks.

**Comment by lafirest** [comment]
@hashforu 
The `iat` only affects the authentication, not authorization(ACL), and authentication is one-time, and then the ACL only check the `exp` value and doesn't care about the `iat` value

**Comment by zmstone** [comment, marked as answer]
> how acl works? check the jwt fileds every pub/pub action? will verify the expired filed the same time? and ,will it cached after first authrization?

JWT ACL result in 'reject' if the time when the check performed is within the token's validity time period. i.e. not before `nbf` and not after `exp`.
The check is performed for each and every publish or subscribe action
However there is a cache which by-default lives 1 minute, so in the worst case, the client might be able to publish or subscribe after `exp` and before cache eviction.

edit: not before`nbf` 
edit2: `iat` has to be in the past (before the time when the check was performed), otherwise result in `issued_in_future` in the error log.

**Comment by idy** [comment]
The `iat` (Issued At) should not be used to detect whether a token is within its valid period. The iat only indicates when the token was created, and there might be slight time differences between different servers, such as a 3-5 second discrepancy.

The `nbf` (Not Before) and `exp` (Expiration Time) should be used to determine whether a token is within its valid period.

**Reply by zmstone** [reply]
> The `iat` (Issued At) should not be used to detect whether a token is within its valid period. The iat only indicates when the token was created, and there might be slight time differences between different servers, such as a 3-5 second discrepancy.

Fair, will consider removing it.

**Reply by zmstone** [reply]
https://github.com/emqx/emqx/pull/13602
